src/citrine/builders/design_spaces.py

```diff
@@ -0,0 +1,96 @@
+import pandas as pd
+from itertools import product
+
+def EnumeratedProductDesignSpaceData(dof_grids:dict):
+    ''' Given a dict of lists or tuples of values for each degree of freedom,
+    create the list of candidates representing the outer product of all values
+    for each degree of freedom
+    
+    Parameters
+    ----------
+    dof_grids: dict(str:list)
+        {'degree of freedom':[list, of, values]} for each DoF
+    '''
+    design_space_tuples = list(product(*dof_grids.values()))
+    design_space_cols = list(dof_grids.keys())
+    df_ds = pd.DataFrame(data=design_space_tuples, columns=design_space_cols)
+    candidates = df_ds.to_dict('records')
+    return candidates
+
+
+def SimpleMixtureProductDesignSpaceData(formulation_grids:dict,
+                                        balance_component:str
+                                        ):
+    ''' Given a dict of lists or tuples of ingredient amounts (must be
+    fractions of some kind, i.e. between 0-1), create
+    a DataFrame that is the outer product of all possible combinations of
+    ingredients *except* for the balance_component. The balance_component
+    will then be filled in as 1-(all other ingredients) for each row.
+    
+    Many points generated through this process will be invalid (negative
+    balance ingredient amount). These rows are filtered out as a last step,
+    if the balance ingredient amount is below the lowest value in its list
```

Comment:
This is a bit confusing to me.  Why include the balance component in the grids if its value is dicated by the others?  How about `balance_minimum: float = 0` and `balance_maximum: float = 1` arguments to make clear that the balance component quantity could lie anywhere between the min and the max?

Reply:
1. You might want to change the balance component, this allows that to be done much more easily
2. It keeps all the ingredients together in a single dict, which just feels nicer
3. You can just pass (min, max) as the dict value for the balance component - fewer arguments

I would rather clarify the docstring than break out the balance component...

Reply:
The ease-of-change is compelling, as is limiting the number of arguments.

Would it be useful to also print a warning message if the balance component maps to a list with the number of elements != 2? Something like "Warning: balance component 'water' mapped to [0.4, 0.5, 0.6], which has more than 2 elements. The min and max values are used to constrain the allowed amount of 'water'; all other values will be ignored."

Reply:
^Yes